### DISCUSSION TITLE
work with API without UI and authentication problem

### DISCUSSION THREAD

**Discussion by loveyandex** [vendurehq/vendure#796] [General]
we need call api with other application other than this UI and browser.
the api is ready but  how the  authentication issue must be solved ?

**Comment by michaelbromley** [comment]
How to authenticate an API call from another program? You can do it in the same way that any other client does - execute the `login` mutation, store the returned cookie/bearer token, and then include it in subsequent requests.

**Reply by gitsupi** [reply]
please give us an example for superadmin username password
and store that cookie token and use it for add a product for example..
how can we catch and use that token in this mutation

**Reply by michaelbromley** [reply]
What have you tried so far?

Here's an example of handling bearer tokens: https://www.vendure.io/docs/storefront/managing-sessions/#bearer-token-sessions

Cookies are similar except they are set with a `Set-Cookie` header.

**Reply by loveyandex** [reply]
how can we catch that token in this mutation?

**Reply by michaelbromley** [reply]
Inspect the headers of the HTTP response.

**Reply by gitsupi** [reply]
an example of login mutation🙏?

**Reply by Izayda** [reply]
A python example:

```
    def login(self, username, password):
        query = """
        mutation {
          login(username: username, password: password, rememberMe: true) {
            ...CurrentUser
            ...ErrorResult
            __typename
          }
        }fragment CurrentUser on CurrentUser {
          id
          identifier
          channels {
            id
            code
            token
            permissions
            __typename
          }
          __typename
        }
        fragment ErrorResult on ErrorResult {
          errorCode
          message
          __typename
        }
        """

        p = requests.post(self.graphql_url, {"query": query})

        if p.status_code == 200:
            self.token_datetime = datetime.datetime.now()
            self.token = p.headers.get('vendure-auth-token')
            return self.token
        else:
            raise Exception
```

**Reply by lusicsmusics1** [reply]
@Izayda  why python example

**Reply by lusicsmusics1** [reply]
> Inspect the headers of the HTTP response.

your docurmnt is very weak 
how can i connect with api

**Reply by michaelbromley** [reply]
@lusicsmusics1 Do you have any experience with connecting to web APIs in general? Vendure is no different really - send a POST request to the server (e.g. `http://localhost:3000/shop-api`) with a GraphQL body. To find out more in particular about graphql, see https://www.howtographql.com/

**Reply by lusicsmusics1** [reply]
@michaelbromley  yes i am 7 languages problem solver
but your design for starting athentication is not really handy
i have no time beacause started a project ..

**Reply by lusicsmusics1** [reply]
@michaelbromley  i think you are never use it before

**Reply by lusicsmusics1** [reply]
`  mutation {
    login(username: "superadmin", password: "superadmin", rememberMe: true) {

    }
  }
`

what should i get in login 
i want save cookies but i can not it in graphql

**Reply by michaelbromley** [reply]
> i want save cookies but i can not it in graphql

If you are making the request from a browser, it should handle cookies automatically (if the server is https, you may need additional config). If not making the request from the browser, I recommend you use the bearer token method of handling the session token. It will be returned as a header from a successful authentication requests, as described here: https://www.vendure.io/docs/storefront/managing-sessions/#bearer-token-sessions

**Reply by lusicsmusics1** [reply]
i am not expert in graphql now nut i need full example to login
no these example are not what i want 
please help me by writing correct mutation request body and ...

**Reply by lusicsmusics1** [reply]
mutation {
    login(username: "superadmin", password: "superadmin", rememberMe: true) {
??
    }
  }

response
{
  "error": {
    "errors": [
      {
        "message": "Cannot query field \"CurrentUser\" on type \"NativeAuthenticationResult\".",
        "locations": [
          {
            "line": 3,
            "column": 5
          }
        ]
      }
    ]
  }
}

**Reply by michaelbromley** [reply]
Ah OK I understand the difficulty you are having now. See this example: https://github.com/vendure-ecommerce/storefront/blob/master/src/app/shared/components/sign-in/sign-in.graphql.ts

The thing with three dots is called "fragment spread syntax" and the reason for it's use in the api is explained in more detail here: https://www.vendure.io/docs/developer-guide/error-handling/

**Reply by lusicsmusics1** [reply]
give me answer use it not an answer that has question self

**Reply by Izayda** [reply]
@michaelbromley you have a lot of patient

**Reply by loveyandex** [reply]
@Izayda you must say he has no even a simple login with his code 😛😛😛

**Reply by loveyandex** [reply]
you are very weak and bad ppl could not present a mutation request

**Reply by michaelbromley** [reply]
Haha, this thread is getting bizarre. OK if anyone else needs professional support, email me via the address on the website and I'll get back to you with my rates 😉
